### PR TITLE
More robust parse and format of overlay details.

### PR DESCRIPTION
**Pull request by tobli**

### CODE REVIEW

**Comment by tobli** [comment]
A bunch of incremental changes to how request/response details are parsed and formatted. Probably makes most sense to review in the commits view to be able to follow what happens.

File: src/ts/waterfall/details-overlay/extract-details-keys.ts [hunk 1]
```diff
@@ -24,6 +32,10 @@ function formatMilliseconds(millis: number): string {
   return `${millis} ms`;
 }
 
+function formatDateLocalized(d: Date): string {
+  return `${d.toUTCString()}</br>(local time: ${d.toLocaleString()})`;
```
Comment (micmro) [inline src/ts/waterfall/details-overlay/extract-details-keys.ts]: Could this cause issues when the recording machine is in a different (non GMT) time-zone than the user looking at the graph? There might be a danger of hiding correlations to backend issues.

File: src/ts/helpers/parse.ts [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+  }
+  return filter(input);
+}
+
+export function formatMilliseconds(millis: number): string {
+  return `${roundNumber(millis, 3)} ms`;
+}
+
+export function formatDateLocalized(date: Date): string {
+  return `${date.toUTCString()}</br>(local time: ${date.toLocaleString()})`;
```
Comment (micmro) [inline src/ts/helpers/parse.ts:69]: Could this cause issues when the recording machine is in a different (non GMT) time-zone than the user looking at the graph? There might be a danger of hiding correlations to backend issues.

Comment (tobli) [reply src/ts/helpers/parse.ts:69]: Yes! =)
It's the same localised date formatting as before, I just extracted it to a named function. The intent was partially to spark a discussion if we should change the logic.

Comment (tobli) [reply src/ts/helpers/parse.ts:69]: When printing it we could use an html5 time element, and some client side js for local formatting.

**Comment by tobli** [comment]
I can post some examples of before/after. I noticed some additional cleanup needed (e.g. displaying empty Redirect URLs). I'll update the PR to cover that today.

**Comment by micmro** [comment]
I noticed with WebPageTest HARs the timings show times for empty `send` and `receive`, which were not shown before.

now:
screen shot 2017-02-07 at 9 58 47 pm
before:
screen shot 2017-02-07 at 9 58 53 pm


Using this HAR [tagesschau.de.160124_HF_0b10bfa565805079d14ed1ae9d79d730.har.zip](https://github.com/micmro/PerfCascade/files/757647/tagesschau.de.160124_HF_0b10bfa565805079d14ed1ae9d79d730.har.zip) (had to zip it to be able to upload it)


Also `Content Size` shows up even though it did not before (might be the same issue than Redirect URLs)

Besides that all looks great.

File: src/ts/helpers/parse.ts (hunk 2 above)
Comment (micmro) [reply src/ts/helpers/parse.ts:69]: Sorry I didn't notice, I thought it was changed because of the `toUTCString ()` 😄

**Comment by tobli** [comment]
Always showing the 0 timings are on purpose, they are mandatory fields in the HAR. A send and receive needs to have happened, even though the har has a 0 number. This is how WPT handles it: https://github.com/WPO-Foundation/webpagetest/blob/master/www/harTiming.inc#L8
"An event that takes 0 milliseconds is considered to have occured.  We assume that it actually took time, but was rounded down."

**Comment by tobli** [comment]
Some other diffs from the sample har (github.com):
before
before

after
after

* ObjectSize is formatted as a byte size
* Treat _cache_time = -1 from WPT as not set (and hide it)
* Hide client port of 0 (which I think WPT does as well).

**Comment by micmro** [comment]
Ok, was not aware of that.

**Comment by tobli** [comment]
Should we merge this now?

**Comment by micmro** [comment]
Sure - Sorry, I thought you'd merge it when you're ready 😃 .

**Comment by tobli** [comment]
Thanks! =)

**Pull request merged**